docs/question/server.md

```diff
@@ -1,17 +1,88 @@
+
+### Step 2: `prepare`
+
+Next, the `prepare` function is called after all elements (e.g. `<pl-number-input>`) have run `generate()`. This is typically done to do any sort of final post-processing, but is not commonly used.
+
+### Step 3: `render`
+
+Next, the `render(data)` function is called to render the question. You can use this function to override how the question is rendered. This is typically only used for more advanced questions.
+
+### Step 4: `parse`
+
+When a student submits their answer, the `parse` function is called to parse the submitted answers after the individual elements have parsed them. This function can be used to display more-specific format errors than the input elements or to parse the input differently. `<pl-number-input>` will have already parsed the submitted value as an integer, and display an error to the student if it was invalid. For this question, we will only allow the student to submit positive integers, so we will check for that and set a format error with `data["format_errors"]` if it is negative.
+
+```python title="server.py"
+def parse(data):
+    # check we don't already have a format error, and if not, check if the answer is negative
+    if "y" not in data["format_errors"] and data["submitted_answers"]["y"] < 0:
+        data["format_errors"]["y"] = "Negative numbers are not allowed"
+```
+
+### Step 5: `grade`
+
+Finally, the `grade(data)` function is called to grade the question. The grade function is responsible for:
+
+- Setting the score and feedback for each named answer in `data["partial_scores"]`.
+- Setting the total score for the question in `data["score"]`.
+- Setting the overall feedback for the question in `data["feedback"]`.
+
+If this function is not defined, the question will be graded automatically based on the correct answers set in `data["correct_answers"]`. Each answer the student provides will also be given feedback.
+
+It is recommended that you give additional feedback to the student beyond right and wrong, as well as partial credit. If the function is defined, the data you receive has already been graded by the elements. You should make sure that you don't override a correct answer by only giving partial credit.
+
+You can also set `data["format_errors"]` to mark the submission as invalid. This will cause the question to not use up one of the student's attempts on assessments.
+
+```python title="server.py"
+import math
+def grade(data):
+    marked_as_incorrect = math.isclose(data["score"], 0.0)
+    if marked_as_incorrect and data["submitted_answers"]["y"] > data["params"]["x"]:
+        data["partial_scores"]["y"]["score"] = 0.5
+        data["score"] = 0.5
+        data["feedback"]["y"] = "Your value for $y$ is larger than $x$, but incorrect."
+```
+
+### Complete example
+
+The finished, complete `question.html` and `server.py` example looks like:
```

Comment:
This example both looks and behaves different than what was built step-by-step above.

Reply:
Which one would you prefer? I did this to reduce complexity in the guide, but the complete example may be more realistic.

Reply:
The main difference between this and the complete example is the randomized operation. I'd be happy to remove that in favor or just having a random parameter and correct answer.

The full example also contains a lot of inline documentation as comments. Would it be feasible to remove some of those (if they've already been discussed elsewhere) or pull them out into normal prose?